packages/@aws-cdk/cdk/lib/app.ts

```diff
@@ -1,20 +1,96 @@
 import cxapi = require('@aws-cdk/cx-api');
-import fs = require('fs');
-import path = require('path');
 import { Stack } from './cloudformation/stack';
-import { IConstruct, MetadataEntry, PATH_SEP, Root } from './core/construct';
+import { Construct } from './core/construct';
+import { IConstruct, MetadataEntry, PATH_SEP } from './core/construct';
+import { Environment } from './environment';
+import { Program } from './program';
+import { validateAndThrow } from './util/validation';
 
 /**
- * Represents a CDK program.
+ * Properties for an App
  */
-export class App extends Root {
+export interface AppProps {
+  /**
+   * The AWS environment (account/region) where stacks in this app will be deployed.
+   *
+   * If not supplied, the `default-account` and `default-region` context parameters will be
+   * used. If they are undefined, it will not be possible to deploy the stack.
+   *
+   * @default Automatically determined
+   */
+  env?: Environment;
+
+  /**
+   * The CDK Program instance in which this app will be defined.
+   *
+   * You don't have to pass this, it exists for testing purposes. Only
+   * supply this parameter if you know what you are doing.
+   *
+   * @default Automatically created
+   */
+  program?: Program;
+}
+
+/**
+ * Used as construct name if no ID is supplied for the App
+ */
+const DEFAULT_APP_NAME = 'App';
+
+/**
+ * An instance of your App
+ */
+export class App extends Construct {
+  /**
+   * True if the given construct is a default-named App
+   */
+  public static isDefaultApp(construct: IConstruct) {
+    return App.isApp(construct) && construct.defaultAppName;
+  }
+
+  /**
+   * True if the given construct is an App object
+   */
+  public static isApp(construct: IConstruct): construct is App {
+    return (construct as any).defaultAppName !== undefined;
```

Comment:
If we were to use Symbols, would that mess with JSII? It would be a safer way to make this determination.

Reply:
I am not sure.

Also, what would the advantage of a symbol be over any instance of an object? Effectively,

```ts
const sym1 = Symbol();
const sym2 = {};
```

`sym1` and `sym2` would behave the same when compared for equality, no?

Reply:
Symbols are unique, therefore there is no way to accidentally add add a 'defaultAppName' property to your construct. They can also be used to compensate for `instanceOf` shortcomings:

```ts
const isApp = Symbol.for('aws-cdk:isApp'); // creates or gets a universal named symbol from the registry

class App {
  [isApp]: true; // will not collide with an ordinary property like 'isApp'
}
```

Reply:
Oh I see, as keys.

Reply:
That will not across JSII, for sure.

Reply:
Does it need to? We only need this deep in the internals to tag App, Stack and Construct.. we can just generate an ordinary property in the JSII facades. It eliminates instanceOf and name-collisions when checking properties.

Reply:
+1 for symbols